src/codec/lz4/encoder.rs

```diff
@@ -0,0 +1,274 @@
+    LZ4F_compressEnd, LZ4F_compressUpdate, LZ4F_createCompressionContext, LZ4F_flush,
+    LZ4F_freeCompressionContext, LZ4F_VERSION,
+};
+
+use crate::{codec::Encode, unshared::Unshared, util::PartialBuffer};
+
+#[derive(Debug)]
+struct EncoderContext {
+    ctx: LZ4FCompressionContext,
+}
+
+#[derive(Debug)]
+enum State {
+    Init {
+        preferences: LZ4FPreferences,
+        buffer: PartialBuffer<Vec<u8>>,
+    },
+    Encoding {
+        buffer: PartialBuffer<Vec<u8>>,
+    },
+    Footer {
+        buffer: PartialBuffer<Vec<u8>>,
+    },
+    Done,
+}
+
+#[derive(Debug)]
+pub struct Lz4Encoder {
+    ctx: Unshared<EncoderContext>,
+    state: State,
+    limit: usize,
+}
+
+impl EncoderContext {
+    fn new() -> Result<Self> {
+        let mut context = LZ4FCompressionContext(core::ptr::null_mut());
+        check_error(unsafe { LZ4F_createCompressionContext(&mut context, LZ4F_VERSION) })?;
+        Ok(Self { ctx: context })
+    }
+}
+
+impl Drop for EncoderContext {
+    fn drop(&mut self) {
+        unsafe { LZ4F_freeCompressionContext(self.ctx) };
+    }
+}
+
+impl Lz4Encoder {
+    pub fn new(level: u32) -> Self {
+        let block_size = BlockSize::Default.get_size();
+        let preferences = LZ4FPreferences {
+            frame_info: LZ4FFrameInfo {
+                block_size_id: BlockSize::Default,
+                block_mode: BlockMode::Linked,
+                content_checksum_flag: ContentChecksum::ChecksumEnabled,
+                content_size: 0,
+                frame_type: FrameType::Frame,
+                dict_id: 0,
+                block_checksum_flag: BlockChecksum::BlockChecksumEnabled,
+            },
+            compression_level: level,
+            auto_flush: 0,
+            favor_dec_speed: 0,
+            reserved: [0; 3],
+        };
+
+        let buffer_size = unsafe { LZ4F_compressBound(block_size, &preferences) };
+
+        Self {
+            ctx: Unshared::new(EncoderContext::new().unwrap()),
+            state: State::Init {
+                preferences,
+                buffer: PartialBuffer::new(Vec::with_capacity(buffer_size)),
+            },
+            limit: block_size,
+        }
+    }
+
+    fn init(
+        ctx: LZ4FCompressionContext,
+        preferences: &LZ4FPreferences,
+        buffer: &mut PartialBuffer<Vec<u8>>,
+    ) -> Result<()> {
+        unsafe {
+            let len = check_error(LZ4F_compressBegin(
+                ctx,
+                buffer.unwritten_mut().as_mut_ptr(),
+                buffer.get_mut().capacity(),
+                preferences,
+            ))?;
+            buffer.get_mut().set_len(len);
+        };
+        Ok(())
+    }
+}
+
+impl Encode for Lz4Encoder {
+    fn encode(
+        &mut self,
+        input: &mut PartialBuffer<impl AsRef<[u8]>>,
+        output: &mut PartialBuffer<impl AsRef<[u8]> + AsMut<[u8]>>,
+    ) -> Result<()> {
+        loop {
+            match &mut self.state {
+                State::Init {
```

Comment:
Can we put init into the new function instead of having a separate state?

Reply:
We probably could. I added it as a separate state since `LZ4F_compressBegin` is fallible. Looking at the [C code](https://github.com/lz4/lz4/blob/9d53d8bb6c4120345a0966e5d8b16d7def1f32c5/lib/lz4frame.c#L690-L814), I see that the only failure it could have is an allocation failure, which should be rare I guess?

Reply:
Yeah...we could have a falliable new as well if people care about it

Reply:
I prefer to keep `new` infallible and have the `Init` state as is if you don't feel strongly about it. If you do, what's your preference out of the 2 options we have:
1. `unwrap()` on the output of `LZ4F_compressBegin`.
2. make `new` infallible.

Reply:
Hmmm thinking it again I think the current is quite good to have error reported instead of panic.